### PR TITLE
.Net: Memory Plugin extraction - Part 2

### PR DESCRIPTION
**Pull request by dmytrostruk**
### Motivation and Context

This PR contains changes to mark Memory-related functionality in Kernel as Obsolete and create alternative methods in `Plugins.Memory` and Connectors projects. 

Note: Obsolete methods and properties will be removed in future releases.

ADR: [0011-memory-as-plugin.md](https://github.com/microsoft/semantic-kernel/blob/main/docs/decisions/0011-memory-as-plugin.md)

**Main change**: helper methods for Memory were moved from `KernelBuilder` to `MemoryBuilder`. Previously, it was possible to configure only one Memory instance per Kernel instance. Now, Memory functionality is decoupled from Kernel and should be configured separately.

Previous Memory usage:
```csharp
var kernel = Kernel.Builder
    .WithOpenAITextEmbeddingGenerationService("text-embedding-ada-002", TestConfiguration.OpenAI.ApiKey)
    .WithMemoryStorage(new AzureCognitiveSearchMemoryStore(TestConfiguration.ACS.Endpoint, TestConfiguration.ACS.ApiKey))
    .Build();

kernel.Memory.SaveInformationAsync(...);
```

New Memory usage:
```csharp
var memory = new MemoryBuilder()
    .WithOpenAITextEmbeddingGenerationService("text-embedding-ada-002", TestConfiguration.OpenAI.ApiKey)
    .WithMemoryStore(new AzureCognitiveSearchMemoryStore(TestConfiguration.ACS.Endpoint, TestConfiguration.ACS.ApiKey))
    .Build();

memory.SaveInformationAsync(...);
```

### Description

1. Added `MemoryBuilder` as equivalent to `KernelBuilder` to initialize the logic related to Memory.
2. Added unit tests for `MemoryBuilder`.
3. Added extension methods for `MemoryBuilder` related to Memory and AI Connectors.
4. Marked Memory-related methods and properties in Kernel/KernelBuilder as obsolete.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Abstractions/IKernel.cs [hunk 1]
```diff
@@ -190,6 +167,19 @@ public interface IKernel
 
     #region Obsolete
 
+    /// <summary>
+    /// Semantic memory instance
+    /// </summary>
+    [Obsolete("Memory functionality will be placed in separate Microsoft.SemanticKernel.Plugins.Memory package. This will be removed in a future release.")]
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/IKernel.cs:173]: We should include some information so a developer know what to do. Can we include a link to a kernel syntax example here?

File: dotnet/src/SemanticKernel.Abstractions/IKernel.cs [hunk 2]
```diff
@@ -201,5 +191,20 @@ public interface IKernel
...
+#pragma warning disable CS1591
+    IDictionary<string, ISKFunction> ImportSkill(object functionsInstance, string? pluginName = null);
+#pragma warning restore CS1591
+
+    /// <summary>
+    /// Set the semantic memory to use
+    /// </summary>
+    /// <param name="memory">Semantic memory instance</param>
+    /// <inheritdoc/>
+    [Obsolete("Memory functionality will be placed in separate Microsoft.SemanticKernel.Plugins.Memory package. This will be removed in a future release.")]
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/IKernel.cs:205]: Again we need to provide sufficient information so a developer knows how to change from the obsolete API.

File: dotnet/src/Plugins/Plugins.Memory/MemoryBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        this._memoryStoreFactory = () => store;
+        return this;
+    }
+
+    /// <summary>
+    /// Add memory store factory.
+    /// </summary>
+    /// <param name="factory">The store factory.</param>
+    /// <returns>Updated Memory builder including the memory store.</returns>
+    public MemoryBuilder WithMemoryStore<TStore>(Func<ILoggerFactory, TStore> factory) where TStore : IMemoryStore
```
Comment (markwallace-microsoft) [inline dotnet/src/Plugins/Plugins.Memory/MemoryBuilder.cs:79]: Should this be named `WithMemoryStoreFactory`?

File: dotnet/samples/KernelSyntaxExamples/Example12_SequentialPlanner.cs [hunk 4]
```diff
@@ -288,24 +289,22 @@ private static IKernel InitializeKernelWithMemory()
...
-    private static ISemanticTextMemory GetMemory(IKernel? kernel = null)
+    private static SemanticTextMemory InitializeMemory()
     {
-        if (kernel is not null)
-        {
-            return kernel.Memory;
-        }
         var memoryStorage = new VolatileMemoryStore();
+
         var textEmbeddingGenerator = new Microsoft.SemanticKernel.Connectors.AI.OpenAI.TextEmbedding.AzureTextEmbeddingGeneration(
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example12_SequentialPlanner.cs:301]: Can we just use `new AzureTextEmbeddingGeneration(`?

**Review by markwallace-microsoft — COMMENTED** [review]
I've added the breaking change tag to this PR as we expect developers to make a code change and this will get the PR listed in the release notes accordingly.

We want to make it easy for developers to understand and adopt the new pattern. Can you:

1. Update the PR description to that includes some instructions on how developers should change their code. It should read like a step by step migration guide. 
1. Add sufficient information to the obsolete message so developers know what they need to change. We don't have a good pattern here but maybe a link to a sample, we may even want dedicated samples that show how to change the code i.e. with before and after sample code.

File: dotnet/src/Plugins/Plugins.Memory/MemoryBuilder.cs (hunk 3 above)
Comment (dmytrostruk) [reply dotnet/src/Plugins/Plugins.Memory/MemoryBuilder.cs:79]: While we have `WithLoggerFactory(ILoggerFactory loggerFactory)`, where we pass `Factory` interface, I don't think it's the case for MemoryStore. The idea here is to create one `WithMemoryStore` method which will have 3 different overloads:
1. `IMemoryStore store` - pass MemoryStore instance as it is.
2. `Func<ILoggerFactory, TStore> factory` - factory for MemoryStore initialization with access to ILoggerFactory.
3. `Func<ILoggerFactory, IDelegatingHandlerFactory, TStore> factory` - factory for MemoryStore initialization with access to ILoggerFactory and IDelegatingHandlerFactory.

Not sure if different name for methods that accept delegate will provide some benefits.

File: dotnet/samples/KernelSyntaxExamples/Example14_SemanticMemory.cs [hunk 5]
```diff
@@ -35,13 +36,19 @@ public static async Task RunAsync()
          * need to worry about embedding generation.
          */
 
-        var kernelWithACS = Kernel.Builder
-            .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+        var kernel = Kernel.Builder
             .WithOpenAITextEmbeddingGenerationService("text-embedding-ada-002", TestConfiguration.OpenAI.ApiKey)
```
Comment (SergeyMenshykh) [inline dotnet/samples/KernelSyntaxExamples/Example14_SemanticMemory.cs:40]: Does Kernel still need the WithOpenAITextEmbeddingGenerationService extension method?

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/samples/KernelSyntaxExamples/Example14_SemanticMemory.cs (hunk 5 above)
Comment (dmytrostruk) [reply dotnet/samples/KernelSyntaxExamples/Example14_SemanticMemory.cs]: In theory Kernel doesn't need it, but it's still possible to register multiple embedding generation services and then get a specific service with `kernel.GetService<ITextEmbeddingGeneration>(name)`. Removing `.GetService` capability from Kernel and related extension methods should be part of separate PR I think.

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**